UpdateUser/handler.ts

```diff
@@ -0,0 +1,148 @@
+import * as express from "express";
+import { toError } from "fp-ts/lib/Either";
+import { identity } from "fp-ts/lib/function";
+import { TaskEither, tryCatch } from "fp-ts/lib/TaskEither";
+import {
+  AzureApiAuthMiddleware,
+  IAzureApiAuthorization,
+  UserGroup
+} from "io-functions-commons/dist/src/utils/middlewares/azure_api_auth";
+import { ContextMiddleware } from "io-functions-commons/dist/src/utils/middlewares/context_middleware";
+import { RequiredBodyPayloadMiddleware } from "io-functions-commons/dist/src/utils/middlewares/required_body_payload";
+import { RequiredParamMiddleware } from "io-functions-commons/dist/src/utils/middlewares/required_param";
+import { withRequestMiddlewares } from "io-functions-commons/dist/src/utils/request_middleware";
+import { wrapRequestHandler } from "italia-ts-commons/lib/request_middleware";
+import {
+  IResponseErrorInternal,
+  IResponseSuccessJson,
+  ResponseSuccessJson
+} from "italia-ts-commons/lib/responses";
+import { NonEmptyString } from "italia-ts-commons/lib/strings";
+import { withoutUndefinedValues } from "italia-ts-commons/lib/types";
+import { EmailAddress } from "../generated/definitions/EmailAddress";
+import { UserCreated } from "../generated/definitions/UserCreated";
+import { UserUpdatePayload } from "../generated/definitions/UserUpdatePayload";
+import {
+  getGraphRbacManagementClient,
+  IServicePrincipalCreds
+} from "../utils/apim";
+import { genericInternalErrorHandler } from "../utils/errorHandler";
+
+type IUpdateUserHandler = (
+  context: Context,
+  auth: IAzureApiAuthorization,
+  email: EmailAddress,
+  userPayload: UserUpdatePayload
+) => Promise<IResponseSuccessJson<UserCreated> | IResponseErrorInternal>;
+
+const getUserFromList = (client: GraphRbacManagementClient, email: string) =>
+  tryCatch(
+    () =>
+      client.users.list({
+        filter: `signInNames/any(x:x/value eq '${email}')`
+      }),
+    toError
+  ).map(userList => userList[0]);
+
+const updateUser = (
+  client: GraphRbacManagementClient,
+  email: EmailAddress,
+  user: User,
+  adb2cTokenAttributeName: string,
+  userPayload: UserUpdatePayload
+): TaskEither<Error, UserCreated> =>
+  tryCatch(
+    () =>
+      client.users.update(
+        user.userPrincipalName,
+        withoutUndefinedValues({
+          displayName:
+            userPayload.first_name && userPayload.last_name
+              ? `${userPayload.first_name} ${userPayload.last_name}`
+              : undefined,
+          givenName: userPayload.first_name,
+          surname: userPayload.last_name,
+          [adb2cTokenAttributeName]: userPayload.token_name
+        })
+      ),
+    toError
+  ).map(
+    updateUserResponse =>
+      ({
+        email,
+        first_name: userPayload.first_name
+          ? userPayload.first_name
+          : user.givenName,
+        id: updateUserResponse.objectId,
+        last_name: userPayload.last_name ? userPayload.last_name : user.surname,
```

Comment:
why this mapping here? can't we just return the exact values ?
first_name -> first_name, last_name -> last_name

Reply:
`UserCreated` has `first_name` and `last_name` as required fields. Since `userPayload` has the same optional attributes, we can map over `user` attributes as a fallback option.

Reply:
we can make it shorter like 

```ts
first_name: userPayload.first_name || user.givenName
```

less cognitive load imho

Reply:
why do we need the fallback?

Reply:
> why do we need the fallback? moreover, it prevent the caller to check if the values are undefined

Because if the caller doesn't provide `first_name` or `last_name` (namely it doesn't want to update first_name or last_name), we must provide the same attributes on `UserCreated` response because they are required. I think that B2C user always has this attributes ('cause createUser expects them in the request payload as required fields)

Reply:
fallback is not a good idea (it's an arbitrary mapping just to fill value to satisfy the decoder). if we must return these values then they must be retrieved from the original user record otherwise we should relax the constraints on these fields

Reply:
Ok let's create a `UserUpdated` instead of `UserCreated` that allows direct mapping from `userPayload`